File: test/zorunlu_alan_baglanti_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:vixrex/config/vitrin_alanlari.g.dart';
import 'package:vixrex/models/store_data.dart';
import 'package:vixrex/services/vixrex_profile_snapshot.dart';

/// Faz F (Tek Asistan planı) "sapma testi" — Dart tarafı.
///
/// PLAN.md: "CI'da sapma testi: iki taraf aynı taslak verisine aynı
/// 'sıradaki alan' cevabını vermeli." Dart ve TypeScript aynı süreçte
/// çalışmadığı için gerçek çapraz-dil çağrısı yapılmıyor; bunun yerine
/// her iki taraf ayrı ayrı, KENDİ şema-bağlantısının eksiksiz olduğunu
/// kanıtlıyor. TS tarafındaki eşleniği:
/// `public_web/tests/zorunlu-alan-baglanti.test.ts`.
///
/// Somut risk: `VixRexProfileSnapshot._alanDolu`'nun switch'i, şemada
/// `zorunlu: true` işaretli ama switch'e case olarak eklenmemiş bir alanı
/// SESSİZCE dolu sayar (`default: return true`). Şemaya yeni zorunlu alan
/// eklenip buraya bağlanmazsa, kurulum akışı o alanı hiç sormaz — tam da
/// il/ilçe'nin bu turda düzeltilen hâli. Bu test, şemadaki HER zorunlu
/// alanın gerçekten ayrı ele alındığını — default fallback'e düşmediğini —
/// kanıtlar.
void main() {
  test("şemadaki her zorunlu alan _alanDolu'da ayrı ele alınır, "
      'sessizce dolu sayılmaz', () {
    final tamStore = StoreData().copyWith(
      name: 'Test Store',
      kategori: 'Kuaför',
      whatsapp: '05551234567',
      address: 'Test Adres',
      provinceName: 'Istanbul',
      districtName: 'Kadikoy',
    );

    for (final alan in zorunluAlanlar) {
      final eksikStore = _tekAlaniBosalt(tamStore, alan.anahtar);
      final snapshot = VixRexProfileSnapshot.from(eksikStore, null);

      expect(
        snapshot.sonrakiEksikZorunluAlan?.anahtar,
        alan.anahtar,
        reason:
            '"${alan.anahtar}" alanı boşken sonrakiEksikZorunluAlan onu '
            'döndürmedi — _alanDolu switch\'inde bu alan için ayrı bir '
            "case yok, default (dolu sayılan) fallback'e düşüyor "
            'olabilir. Şemaya yeni zorunlu alan eklerken '
            '_alanDolu ve nextMissingField switch\'lerine de case '
            'eklenmeli.',
      );
    }
  });
}

/// Diğer her şey doluyken yalnız [anahtar]'ı boşaltır.
StoreData _tekAlaniBosalt(StoreData dolu, String anahtar) {
  switch (anahtar) {
    case 'isletmeAdi':
      return dolu.copyWith(name: '');
    case 'kategori':
      return dolu.copyWith(kategori: '');
    case 'whatsapp':
      return dolu.copyWith(whatsapp: '');
    case 'adres':
      return dolu.copyWith(address: '');
    case 'il':
      return dolu.copyWith(provinceName: '');
    case 'ilce':
      return dolu.copyWith(districtName: '');
    default:
      throw StateError(
        '"$anahtar" için _tekAlaniBosalt bilmiyor — bu test dosyası yeni '
        'zorunlu alanla birlikte güncellenmeli (bkz. dosya başı notu: '
        'sessiz "dolu sayma" riskini bu test yakalar, ama yeni alanı '
        "burada boşaltmayı bilmezse testin kendisi anlamsızlaşır).",
      );
  }
}

File: public_web/tests/zorunlu-alan-baglanti.test.ts
import { describe, expect, it } from "vitest";
import { hazirlikRaporu } from "../src/lib/vitrinReadiness";
import { VITRIN_FIELDS } from "../src/lib/vitrinFieldSchema";

/**
 * Faz F (Tek Asistan planı) "sapma testi" — Next.js tarafı.
 *
 * PLAN.md: "CI'da sapma testi: iki taraf aynı taslak verisine aynı
 * 'sıradaki alan' cevabını vermeli." Dart ve TypeScript aynı süreçte
 * çalışmadığı için gerçek çapraz-dil çağrısı yapılmıyor; her iki taraf
 * ayrı ayrı KENDİ şema-bağlantısının eksiksiz olduğunu kanıtlıyor. Dart
 * tarafındaki eşleniği: test/zorunlu_alan_baglanti_test.dart.
 *
 * Next.js tarafında `doluMu()` şemadan JENERİK okuduğu için (Dart'ın
 * elle yazılmış `_alanDolu` switch'inin aksine) "yeni alan eklenip
 * sessizce dolu sayılma" riski yapısal olarak yok — ama bu test yine de
 * her zorunlu alanın gerçekten TEK BAŞINA eksikken doğru tespit
 * edildiğini kanıtlar. Asıl amaç: iki taraf aynı zorunlu alan kümesini
 * aynı şekilde yorumluyor mu, ampirik kanıt.
 */
describe("şemadaki her zorunlu alan hazirlikRaporu'da ayrı ele alınır", () => {
  const zorunluAlanlar = VITRIN_FIELDS.filter((a) => a.zorunlu);

  it("en az bir zorunlu alan var (kurulum akışı boşsa test anlamsız)", () => {
    expect(zorunluAlanlar.length).toBeGreaterThan(0);
  });

  for (const eksikAlan of zorunluAlanlar) {
    it(`yalnız "${eksikAlan.anahtar}" boşken eksikler listesinde tek başına görünür`, () => {
      const draft: Record<string, unknown> = {};
      for (const alan of zorunluAlanlar) {
        if (alan.anahtar === eksikAlan.anahtar) continue;
        draft[alan.kolon] = "dolu";
      }

      const rapor = hazirlikRaporu(draft);
      const eksikAnahtarlar = rapor.eksikler
        .filter((e) => e.onem === "temel")
        .map((e) => e.anahtar);

      expect(
        eksikAnahtarlar,
        `"${eksikAlan.anahtar}" boşken temel eksikler listesi ${JSON.stringify(
          eksikAnahtarlar,
        )} — beklenen: ["${eksikAlan.anahtar}"]. alanOnemi()/TEMEL_ALANLAR ` +
          "bu alanı doğru sınıflandırmıyor olabilir.",
      ).toEqual([eksikAlan.anahtar]);
    });
  }
});
